mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,306 @@
+        try:
+            output = subprocess.check_output(path_to_version_sh)
+        except OSError as e:
+            warn("Could not record emirge's package versions: " + str(e))
+
+    mgr.set_constant("emirge_package_versions", output)
+
+# }}}
+
+
+# {{{ State handling
+
+def set_sim_state(mgr: LogManager, dim, state, eos) -> None:
+    """Update the simulation state of all :class:`StateConsumer` of the log manager.
+
+    Parameters
+    ----------
+    mgr
+        The :class:`logpyle.LogManager` to set the state of.
+    """
+    for gd_lst in [mgr.before_gather_descriptors,
+            mgr.after_gather_descriptors]:
+        for gd in gd_lst:
+            if isinstance(gd.quantity, StateConsumer):
+                gd.quantity.set_sim_state(dim, state, eos)
+
+
+class StateConsumer:
+    """Base class for quantities that require a state for logging."""
+
+    def __init__(self):
+        self.state_vars = None
+
+    def set_sim_state(self, dim, state, eos) -> None:
+        """Update the state vector of the object."""
+        self.state_vars = extract_state_vars(dim, state, eos)
+
+# }}}
+
+# {{{ Discretization-based quantities
+
+
+class DiscretizationBasedQuantity(LogQuantity, StateConsumer):
+    """Logging support for physical quantities.
+
+    Possible rank aggregation operations (`op`) are: min, max, norm.
+    """
+
+    def __init__(self, discr: Discretization, quantity: str, op: str,
+                 unit: str = None, name: str = None, dim: int = None):
+        if unit is None:
+            unit = units_logging(quantity)
+
+        if name is None:
+            name = f"{op}_{quantity}" + (str(dim) if dim is not None else "")
+
+        LogQuantity.__init__(self, name, unit)
+        StateConsumer.__init__(self)
+
+        self.discr = discr
+
+        self.quantity = quantity
+        self.dim = dim
+
+        from functools import partial
+
+        if op not in ["min", "max", "norm"]:
+            raise ValueError("op must be one of 'min', 'max', 'norm'.")
+
+        if op == "min":
+            self._discr_reduction = partial(self.discr.nodal_min, "vol")
+            self.rank_aggr = min
+        elif op == "max":
+            self._discr_reduction = partial(self.discr.nodal_max, "vol")
+            self.rank_aggr = max
+        elif op == "norm":
+            self._discr_reduction = partial(self.discr.norm)
+            self.rank_aggr = max
+        else:
+            raise RuntimeError(f"unknown operation {op}")
+
+    @property
+    def default_aggregator(self):
+        """Rank aggregator to use."""
+        return self.rank_aggr
+
+    def __call__(self):
+        """Return the requested quantity."""
+        if self.state_vars is None:
+            return None
+
+        quantity = self.state_vars[self.quantity]
+
+        if self.dim is not None:  # momentum
+            return self._discr_reduction(quantity[self.dim])
+        else:  # mass, energy
+            return self._discr_reduction(quantity)
```

Comment:
Will this break when we have other non-scalar quantities (e.g. `scalar_mass` for mixtures) who's dimension is not spatially-dependent?

Reply:
I'm not sure I understood correctly, but I think this will not break. In this context `self.dim` isn't necessarily related or restricted to the number of dimensions of the input problem. It might be better to rename that variable to something like `q_dim` or so.

Reply:
Yup, perfect!  Thanks!